### ISSUE TITLE
Array type inference with local function

### ISSUE THREAD

**Issue by RblSb** [HaxeFoundation/haxe#9987]
Seems pretty minor, but maybe report is useful (Haxe nightly)
```haxe
class Point {
	public function new() {}
	public var x = 0;
	public var y = 0;
}
class Main {
	static function main() {
		final points = [];
		function foo() points[0].y = 1;
		points[0] = new Point();
		points[0].x = 1; // { y : Int } has no field x
		foo();
	}
}
```
Same result with field and function before main:
```haxe
static final points = [];
static function foo() points[0].y = 1;
static function main() {...
```

**Comment by nadako** [comment]
I believe this is related to the fact that AFAIK monomorphs (unknown types) are "closed" at function exit, but I wonder if we should close monomorphs that were created outside of the function we're processing?

Maybe it might be that I have no idea what I'm talking about, so let's wait for @Simn :)

**Comment by Simn** [comment]
Hmm, interesting. I think nadako's analysis is correct, which suggests some state isn't managed correctly with regards to local functions.

**Comment by Simn** [comment]
Right, so while `type_function` does have `ctx.monomorphs.perfunction <- []`, the issue here is that upon field access on `CUnknown` we add the accessed monomorph back (in `type_field`). I'm now looking at this wondering why we do that. All monomorphs that we're interested in should already live there, and accessing their structure should not affect how and when they are closed...

Disabling this breaks this test from #6259:

```haxe
class Main {
	static function main() {
		function f(a)
			return a.b;
		var res = f({b: (s) -> s.length})("6259");
		var res = f({b: (s) -> s.length})({length: 10}); // { length : Int } should be String
	}
}
```

This is some weird ass corner case about the return type of `f`...
